File: examples/csharp/data-types/DataTypes.cs
// 第 09 章 · 数据类型 — C# 示例
// 运行：dotnet new console -o app，复制为 app/Program.cs，再 dotnet run --project app
using System;

class DataTypes
{
    static void Main()
    {
        // 1. 宽度固定
        Console.WriteLine($"int 范围: {int.MinValue} ~ {int.MaxValue}");
        Console.WriteLine($"宽度(字节): int={sizeof(int)} long={sizeof(long)} decimal={sizeof(decimal)}");

        // 2. 浮点误差
        Console.WriteLine($"0.1 + 0.2 = {0.1 + 0.2} | 等于 0.3 吗: {0.1 + 0.2 == 0.3}");

        // 3. decimal 是语言原生的精确十进制（注意 m 后缀）
        Console.WriteLine($"0.1m + 0.2m = {0.1m + 0.2m} | 等于 0.3m 吗: {0.1m + 0.2m == 0.3m}");

        // 4. 溢出：默认静默回绕，checked 可捕获
        int max = int.MaxValue;
        unchecked { Console.WriteLine($"unchecked 最大值+1 = {max + 1}  ← 回绕"); }
        try { checked { int bad = max + 1; } }
        catch (OverflowException) { Console.WriteLine("checked 块中会抛 OverflowException"); }

        // 5. 字符串长度数的是 UTF-16 码元
        string wave = "👋";
        Console.WriteLine($"\"👋\".Length = {wave.Length} (UTF-16 码元)");
    }
}

File: examples/java/data-types/DataTypes.java
// 第 09 章 · 数据类型 — Java 示例
// 运行：javac DataTypes.java && java DataTypes
import java.math.BigDecimal;

public class DataTypes {
    public static void main(String[] args) {
        // 1. 基本类型宽度由规范固定
        System.out.println("int 范围: " + Integer.MIN_VALUE + " ~ " + Integer.MAX_VALUE);

        // 2. 溢出静默回绕
        int max = Integer.MAX_VALUE;
        System.out.println("最大值 + 1 = " + (max + 1) + "  ← 变负数了，且不报错");
        try {
            Math.addExact(max, 1);
        } catch (ArithmeticException e) {
            System.out.println("用 Math.addExact 才会抛异常: " + e.getMessage());
        }

        // 3. 浮点误差
        System.out.println("0.1 + 0.2 = " + (0.1 + 0.2) + " | 等于 0.3 吗: " + (0.1 + 0.2 == 0.3));

        // 4. 金额用 BigDecimal（从字符串构造）
        BigDecimal a = new BigDecimal("0.1");
        BigDecimal b = new BigDecimal("0.2");
        System.out.println("BigDecimal 精确: " + a.add(b)
            + " | 等于 0.3 吗: " + (a.add(b).compareTo(new BigDecimal("0.3")) == 0));

        // 5. 字符串长度数的是 UTF-16 码元
        String wave = "👋";
        System.out.println("length() = " + wave.length()
            + " | codePointCount = " + wave.codePointCount(0, wave.length()));
    }
}
